File: py/dockit/generadores/guion.py
"""El guion: describe un documento sin comprometerse con el formato.

El mismo guion produce el informe, las diapositivas y la hoja de cálculo. Cada
generador lo interpreta a su manera, pero la validación es una sola y vive aquí.
"""
from __future__ import annotations

CLASES = {"titulo", "parrafo", "lista", "tabla", "figura", "cita",
          "salto", "bibliografia"}
TIPOS = {"docx", "pptx", "xlsx", "md"}


class GuionInvalido(ValueError):
    """El guion no describe un documento que se pueda producir."""


def _validar_bloque(b: dict, i: int) -> None:
    clase = b.get("clase")
    donde = f"bloque {i + 1}"
    if clase not in CLASES:
        raise GuionInvalido(f"{donde}: clase de bloque desconocida: {clase!r}")

    if clase == "titulo":
        if not b.get("texto"):
            raise GuionInvalido(f"{donde}: un título sin texto no sirve de nada")
        nivel = b.get("nivel", 0)
        if not isinstance(nivel, int) or not 1 <= nivel <= 6:
            raise GuionInvalido(f"{donde}: nivel de título fuera de rango: {nivel}")

    elif clase in ("parrafo", "cita"):
        if not b.get("texto"):
            raise GuionInvalido(f"{donde}: un bloque {clase} necesita texto")

    elif clase == "lista":
        if not b.get("items"):
            raise GuionInvalido(f"{donde}: una lista sin elementos")

    elif clase == "tabla":
        filas = b.get("filas") or []
        if not filas:
            raise GuionInvalido(f"{donde}: una tabla sin filas")
        cab = b.get("cabecera") or []
        if cab:
            for n, fila in enumerate(filas, 1):
                if len(fila) != len(cab):
                    raise GuionInvalido(
                        f"{donde}: la fila {n} tiene {len(fila)} celdas "
                        f"y la cabecera {len(cab)}")

    elif clase == "figura":
        if not b.get("ruta"):
            raise GuionInvalido(f"{donde}: una figura sin ruta de imagen")


def validar(guion: dict, disponibles: set[str] | None = None) -> None:
    """Comprueba el guion entero.

    `disponibles` son las claves de la bibliografía. Citar algo que no está ahí
    es exactamente el error que produce una referencia inventada, así que se
    rechaza al producir, no solo al verificar.
    """
    if guion.get("tipo") not in TIPOS:
        raise GuionInvalido(f"tipo de salida desconocido: {guion.get('tipo')!r}")
    if not guion.get("titulo"):
        raise GuionInvalido("el guion no tiene título")
    bloques = guion.get("bloques") or []
    if not bloques:
        raise GuionInvalido("el guion no tiene bloques")

    for i, b in enumerate(bloques):
        _validar_bloque(b, i)
        if disponibles is not None:
            for c in b.get("citas") or []:
                if c not in disponibles:
                    raise GuionInvalido(
                        f"bloque {i + 1} cita «{c}», que no está en la bibliografía")


def texto_con_citas(bloque: dict, en_texto: dict[str, str]) -> str:
    """Pega las citas al final de la frase, como manda APA.

    Se usa el texto que produjo citeproc, nunca uno escrito a mano: si no,
    el «et al.» y el orden de autores se desincronizan con la bibliografía.
    """
    texto = bloque.get("texto", "")
    claves = bloque.get("citas") or []
    if not claves:
        return texto
    citas = " ".join(en_texto.get(c, f"({c})") for c in claves)
    cuerpo = texto.rstrip()
    punto = cuerpo.endswith((".", "!", "?"))
    if punto:
        cuerpo = cuerpo[:-1].rstrip()
    return f"{cuerpo} {citas}." if punto else f"{cuerpo} {citas}"

File: internal/dominio/guion.go
package dominio

import (
	"errors"
	"fmt"
)

// ClaseBloque es cada cosa que puede aparecer en un documento.
type ClaseBloque string

const (
	Titulo       ClaseBloque = "titulo"
	Parrafo      ClaseBloque = "parrafo"
	Lista        ClaseBloque = "lista"
	Tabla        ClaseBloque = "tabla"
	Figura       ClaseBloque = "figura"
	Cita         ClaseBloque = "cita"
	SaltoPagina  ClaseBloque = "salto"
	Bibliografia ClaseBloque = "bibliografia"
)

var clasesValidas = map[ClaseBloque]bool{
	Titulo: true, Parrafo: true, Lista: true, Tabla: true,
	Figura: true, Cita: true, SaltoPagina: true, Bibliografia: true,
}

var tiposSalida = map[string]bool{"docx": true, "pptx": true, "xlsx": true, "md": true}

// Bloque es una unidad del documento, independiente del formato final.
type Bloque struct {
	Clase    ClaseBloque `json:"clase"`
	Nivel    int         `json:"nivel,omitempty"`
	Texto    string      `json:"texto,omitempty"`
	Items    []string    `json:"items,omitempty"`
	Cabecera []string    `json:"cabecera,omitempty"`
	Filas    [][]string  `json:"filas,omitempty"`
	Ruta     string      `json:"ruta,omitempty"`
	Leyenda  string      `json:"leyenda,omitempty"`
	Fuente   string      `json:"fuente,omitempty"`
	Citas    []string    `json:"citas,omitempty"`
	Notas    string      `json:"notas,omitempty"`
}

func (b Bloque) Validar() error {
	if !clasesValidas[b.Clase] {
		return fmt.Errorf("clase de bloque desconocida: %q", b.Clase)
	}
	switch b.Clase {
	case Titulo:
		if b.Texto == "" {
			return errors.New("un título sin texto no sirve de nada")
		}
		if b.Nivel < 1 || b.Nivel > 6 {
			return fmt.Errorf("nivel de título fuera de rango: %d", b.Nivel)
		}
	case Parrafo, Cita:
		if b.Texto == "" {
			return fmt.Errorf("un bloque %s necesita texto", b.Clase)
		}
	case Lista:
		if len(b.Items) == 0 {
			return errors.New("una lista sin elementos")
		}
	case Tabla:
		if len(b.Filas) == 0 {
			return errors.New("una tabla sin filas")
		}
		if n := len(b.Cabecera); n > 0 {
			for i, f := range b.Filas {
				if len(f) != n {
					return fmt.Errorf("la fila %d tiene %d celdas y la cabecera %d",
						i+1, len(f), n)
				}
			}
		}
	case Figura:
		if b.Ruta == "" {
			return errors.New("una figura sin ruta de imagen")
		}
	}
	return nil
}

// Guion describe un documento sin comprometerse con el formato: el mismo guion
// produce el informe, las diapositivas y la hoja de cálculo.
type Guion struct {
	Tipo    string   `json:"tipo"`
	Titulo  string   `json:"titulo"`
	Autor   string   `json:"autor,omitempty"`
	Bloques []Bloque `json:"bloques"`
}

// Validar comprueba el guion entero. `disponibles` son las claves de la
// bibliografía: citar algo que no está ahí es el error que produce una
// referencia inventada, así que se rechaza aquí.
func (g Guion) Validar(disponibles map[string]bool) error {
	if !tiposSalida[g.Tipo] {
		return fmt.Errorf("tipo de salida desconocido: %q", g.Tipo)
	}
	if g.Titulo == "" {
		return errors.New("el guion no tiene título")
	}
	if len(g.Bloques) == 0 {
		return errors.New("el guion no tiene bloques")
	}
	for i, b := range g.Bloques {
		if err := b.Validar(); err != nil {
			return fmt.Errorf("bloque %d: %w", i+1, err)
		}
		for _, c := range b.Citas {
			if disponibles != nil && !disponibles[c] {
				return fmt.Errorf("bloque %d cita «%s», que no está en la bibliografía", i+1, c)
			}
		}
	}
	return nil
}
